### PR TITLE
Make publish command support bump-version types (patch, minor, major) and support bumping Dockerfile for you

### PR DESCRIPTION
**Pull request by noahkawasaki-airbyte**
## What
Applying some of the relevant refactors from https://github.com/airbytehq/airbyte/pull/12474/files into the current `/publish` slash command.

## How
Includes:

1) Changing `auto-bump-version` boolean input to `bump-version` string input for patch, minor, or major.
2) Adding the Dockerfile into the files which are auto-bumped for you, and basing it off of master and not the current (potentially stale) branch
3) Split the build and test step out of the publish step
4) Some slight re-ordering of events in the workflow
5) Harden up the condition for when a bump occurred to be based on an action output determined dynamically from the Bump step. This means that this step is capable of correctly telling the rest of the workflow whether or not a bump occurred even if someone continues to run the publish command over and over again or if a merge conflict occurs. It should always correctly determine whether a version change happened more reliably

## Recommended reading order
1. manage.sh
2. publish-command.yml

### CODE REVIEW

File: .github/workflows/publish-command.yml [hunk 1]
```diff
@@ -195,57 +249,6 @@ jobs:
...
-          sourcename=$(yq e ".[] | select(has(\"dockerRepository\")) | select(.dockerRepository == \"$connector\") | .name" "$definitionpath"source_definitions.yaml)
-          destname=$(yq e ".[] | select(has(\"dockerRepository\")) | select(.dockerRepository == \"$connector\") | .name" "$definitionpath"destination_definitions.yaml)
-          if [ -z "$sourcename" ]
-          then yq e "(.[] | select(.name == \"$destname\").dockerImageTag)|=\"${{ env.IMAGE_VERSION }}\"" -i "$definitionpath"destination_definitions.yaml
-          else yq e "(.[] | select(.name == \"$sourcename\").dockerImageTag)|=\"${{ env.IMAGE_VERSION }}\"" -i "$definitionpath"source_definitions.yaml
-          fi
-      - name: Run gradle process changes
-        if: github.event.inputs.auto-bump-version == 'true' && success()
-        run: |
-          ./gradlew :airbyte-config:init:processResources
```
Comment (noahkawasaki-airbyte) [inline .github/workflows/publish-command.yml:222]: All of this has been moved into manage.sh

File: tools/integrations/manage.sh [hunk 2]
```diff
@@ -135,39 +114,121 @@ cmd_bump_version() {
...
-  _error_if_tag_exists "$image_name:$bumped_version"
-  echo "$connector:$current_version will be bumped to $connector:$bumped_version"
+  if [[ "$FORCE_BUMP" == "true" ]] || [[ "$bumped_version" != "$master_version" && "$bumped_version" != "$branch_version" ]]; then
+    _error_if_tag_exists "$image_name:$bumped_version"
+    echo "$connector:$branch_version will be bumped to $connector:$bumped_version"
+
+    # Set outputs back to Github Actions for later steps
+    echo ::set-output name=master_version::"${master_version}"
+    echo ::set-output name=bumped_version::"${bumped_version}"
+    echo ::set-output name=bumped::"true"
```
Comment (noahkawasaki-airbyte) [inline tools/integrations/manage.sh:128]: This GHA output is important, its what it used to tell downstream steps whether or not a bump occurred

Comment (noahkawasaki-airbyte) [reply tools/integrations/manage.sh:128]: Reason this is better is for future runs when a bump has already occurred, future GHA runs will not fail or run unnecessarily

File: tools/integrations/manage.sh [hunk 3]
```diff
@@ -208,12 +269,6 @@ cmd_publish() {
   # time building, running tests to realize this version is a duplicate.
   _error_if_tag_exists "$versioned_image"
 
-  # building the connector
-  cmd_build "$path" "$run_tests"
-
-  # in case curing the build / tests someone this version has been published.
-  _error_if_tag_exists "$versioned_image"
-
```
Comment (noahkawasaki-airbyte) [inline tools/integrations/manage.sh:216]: These have been pulled out into different steps cmd_build, cmd_test, both of which have separate GHA steps too. Makes it easier to see where a failure happens

File: .github/workflows/publish-command.yml [hunk 4]
```diff
@@ -128,45 +135,92 @@ jobs:
...
+        with:
+          comment-id: ${{ github.event.inputs.comment-id }}
+          body: |
+            > :x: Couldn't auto-bump version for ${{github.event.inputs.connector}}
+      - name: Build ${{ github.event.inputs.connector }}
         run: |
-          source tools/lib/lib.sh
-          DOCKERFILE=airbyte-integrations/${{ github.event.inputs.connector }}/Dockerfile
-          echo "IMAGE_NAME=$(echo ${{ github.event.inputs.connector }} | cut -d"/" -f2)" >> $GITHUB_ENV
-          echo "IMAGE_VERSION=$(_get_docker_image_version ${DOCKERFILE})" >> $GITHUB_ENV
```
Comment (noahkawasaki-airbyte) [inline .github/workflows/publish-command.yml:143]: So a big-ish change happening is that this workflow now bumps the dockerfile for you and handles the version string as an output from the previously done Bump step

File: .github/workflows/publish-command.yml [hunk 5]
```diff
@@ -128,45 +135,92 @@ jobs:
...
+        id: build
+        env:
+          DOCKER_PASSWORD: ${{ secrets.DOCKER_PASSWORD }}
+          # Oracle expects this variable to be set. Although usually present, this is not set by default on GitHub virtual runners.
+          TZ: UTC
+      - name: Test ${{ github.event.inputs.connector }}
+        if: github.event.inputs.run-tests == 'true'
+        run: |
+          ./tools/integrations/manage.sh test airbyte-integrations/${{ github.event.inputs.connector }}
       - name: Prepare Sentry
```
Comment (noahkawasaki-airbyte) [inline .github/workflows/publish-command.yml:181]: I'm not super familiar with what a "Sentry Release" is, but I'm wondering if this stuff should all happen after the publish step runs and succeeds. Would we want a sentry release to be created before a publish step that coul potentially fail?

File: .github/workflows/publish-command.yml [hunk 6]
```diff
@@ -128,45 +135,92 @@ jobs:
...
         id: publish
         env:
           DOCKER_USERNAME: airbytebot
           DOCKER_PASSWORD: ${{ secrets.DOCKER_PASSWORD }}
           # Oracle expects this variable to be set. Although usually present, this is not set by default on Github virtual runners.
           TZ: UTC
+      - name: Process Build ${{ github.event.inputs.connector }}
+        run: |
+          export publish_spec_to_cache=true
+          ./tools/integrations/manage.sh process_build airbyte-integrations/${{ github.event.inputs.connector }}
```
Comment (noahkawasaki-airbyte) [inline .github/workflows/publish-command.yml:206]: This basically handles the publish spec to cache and gradle processResources generator steps

File: tools/integrations/manage.sh [hunk 7]
```diff
@@ -43,57 +43,26 @@ cmd_build() {
   echo "Building $path"
   ./gradlew --no-daemon "$(_to_gradle_path "$path" clean)"
   ./gradlew --no-daemon "$(_to_gradle_path "$path" build)"
-
-  if [ "$run_tests" = false ] ; then
```
Comment (noahkawasaki-airbyte) [inline tools/integrations/manage.sh:47]: test running has been moved into its own step

File: tools/integrations/manage.sh [hunk 8]
```diff
@@ -226,49 +281,11 @@ cmd_publish() {
     docker tag "$image_name:dev" "$latest_image"
 
     echo "Publishing new version ($versioned_image)"
-    docker push "$versioned_image"
-    docker push "$latest_image"
+#    docker push "$versioned_image"
+#    docker push "$latest_image"
```
Comment (noahkawasaki-airbyte) [inline tools/integrations/manage.sh:285]: Commented out intentionally so I dont push new images while iterating

File: .github/workflows/publish-command.yml (hunk 5 above)
Comment (evantahler) [reply .github/workflows/publish-command.yml:182]: I also don't really know what we are using sentry for - @Phlair have you come across any data in sentry lately?  Maybe we can remove this

**Review by evantahler — APPROVED** [review]
I like this direction a lot. 
 
To clarify, when this is merged in:  As a connector developer releasing a new version, I'll have one slash command now that will both bump my version and publish the new docker image.  
  1. I think this is still to be done on my branch, before merging to master.  I think this is OK for now, and I think we are working towards this happening automatically when the connector branch is merged to master perhaps?
  2. I think that I'll still manually need to write the changelog for this version in the doc.  When in the workflow should I do that?

**Comment by noahkawasaki-airbyte** [comment]
Yep so:

1) This is also how it currently works, this PR mostly changes the version bumping to accept a bump-version _type_ (patch, minor, major) and bumps the Dockerfile for you. It is still weird that this happens on this same branch before merging because the publish still occurs before merge. And the ideal end result is the publishing stuff actually occurs after a PR is merged but thats out of the scope of this particular PR.
2) Changelog.. unfortunate but requires manual update whenever. So yeah, author needs to be aware of the version getting bumped to and make the relevant updates until some automation is added there

**Comment by evantahler** [comment]
The sources-dx team has inherited this story from the sources-databases team.

**Pull request closed without merging**